### PR TITLE
[no squash] VoxelArea fixes v2

### LINKED ISSUE #15516: Server crash due to assertion failure in voxel.h
### Luanti version

```true
Luanti 5.10.0-debug (Linux)
Using LuaJIT 2.1.0-beta3
Built by GCC 9.4
Running on Linux/5.15.0 x86_64
BUILD_TYPE=Debug
RUN_IN_PLACE=1
USE_CURL=1
STATIC_SHAREDIR="."
```

### Irrlicht device

_No response_

### Operating system and version

Ubuntu 20.04.6 LTS

### CPU model

_No response_

### GPU model

_No response_

### Active renderer

_No response_

### Summary

Hello,
I've been encountering random crashes on my Luanti server. The crash happens without any prior warning or error messages in the logs. I managed to generate a core dump during one of the crashes. Below is the backtrace obtained using gdb:
```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f29b06b8859 in __GI_abort () at abort.c:79
#2  0x00007f29b06b8729 in __assert_fail_base (fmt=0x7f29ac13126f <error: Cannot access memory at address 0x7f29ac13126f>, assertion=0x55e8b1db2fdc "m_cache_extent.X >= 0", 
    file=0x55e8b1d65728 "/home/ubuntu/minetest/src/voxel.h", line=335, function=<optimized out>) at assert.c:92
#3  0x00007f29b06c9fd6 in __GI___assert_fail (assertion=0x55e8b1db2fdc "m_cache_extent.X >= 0", file=0x55e8b1d65728 "/home/ubuntu/minetest/src/voxel.h", line=335, 
    function=0x55e8b1db2fbe "void VoxelArea::cacheExtent()") at assert.c:101
#4  0x000055e8b17e89dd in VoxelArea::cacheExtent (this=0x7f299f4ec700) at /home/ubuntu/minetest/src/voxel.h:335
#5  0x000055e8b1ac65e2 in VoxelArea::addPoint (this=0x7f299f4ec700, p=...) at /home/ubuntu/minetest/src/voxel.h:98
#6  0x000055e8b1ac67d8 in MapEditEvent::getArea (this=0x7f299f4ec9c0) at /home/ubuntu/minetest/src/map.h:83
#7  0x000055e8b1ad11a0 in Server::onMapEditEvent (this=0x7fff17ce2b30, event=...) at /home/ubuntu/minetest/src/server.cpp:1313
#8  0x000055e8b1a87661 in Map::dispatchEvent (this=0x55e8b81bca20, event=...) at /home/ubuntu/minetest/src/map.cpp:54
#9  0x000055e8b1acda35 in Server::AsyncRunStep (this=0x7fff17ce2b30, dtime=0.106031999, initial_step=false) at /home/ubuntu/minetest/src/server.cpp:745
#10 0x000055e8b1ac79c9 in ServerThread::run (this=0x55e8b817c2c0) at /home/ubuntu/minetest/src/server.cpp:139
#11 0x000055e8b1c30403 in Thread::threadProc (thr=0x55e8b817c2c0) at /home/ubuntu/minetest/src/threading/thread.cpp:194
#12 0x000055e8b1c31086 in std::__invoke_impl<void, void (*)(Thread*), Thread*> (__f=@0x55e8bb92d540: 0x55e8b1c30358 <Thread::threadProc(Thread*)>) at /usr/include/c++/9/bits/invoke.h:60
#13 0x000055e8b1c30ffb in std::__invoke<void (*)(Thread*), Thread*> (__fn=@0x55e8bb92d540: 0x55e8b1c30358 <Thread::threadProc(Thread*)>) at /usr/include/c++/9/bits/invoke.h:95
#14 0x000055e8b1c30f6b in std::thread::_Invoker<std::tuple<void (*)(Thread*), Thread*> >::_M_invoke<0ul, 1ul> (this=0x55e8bb92d538) at /usr/include/c++/9/thread:244
#15 0x000055e8b1c30f26 in std::thread::_Invoker<std::tuple<void (*)(Thread*), Thread*> >::operator() (this=0x55e8bb92d538) at /usr/include/c++/9/thread:251
#16 0x000055e8b1c30f0c in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(Thread*), Thread*> > >::_M_run (this=0x55e8bb92d530) at /usr/include/c++/9/thread:195
#17 0x00007f29b099cdf4 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#18 0x00007f29b0890609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#19 0x00007f29b07b5353 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

### Steps to reproduce

I have absolutely no idea how this bug occurs but the server crashed 8 times in 20 minutes for this error.

### PR DESCRIPTION
**Pull request by sfan5**
fixes #15516 and other edge cases that haven't caused problems *yet*

## To do

This PR is Ready for Review.

## How to test

1. observe unit test results
2. game should work normally

### CODE REVIEW

File: src/voxel.h [hunk 1]
```diff
@@ -109,21 +109,20 @@ class VoxelArea
...
-		// FIXME: shouldn't this actually be a component-wise check?
-		return m_cache_extent == v3s16(0,0,0);
+		return !m_cache_extent.X || !m_cache_extent.Y || !m_cache_extent.Z;
 	}
 
-	s32 getVolume() const
+	u32 getVolume() const
 	{
-		return (s32)m_cache_extent.X * (s32)m_cache_extent.Y * (s32)m_cache_extent.Z;
+		// FIXME: possible integer overflow here
```
Comment (SmallJoker) [inline src/voxel.h:124]: 16+16+16 = 48 bits. Thus, how about using `u64` here?

Comment (sfan5) [reply src/voxel.h:124]: that would work but I thought "well we don't really work with areas this big so I can get away with not fixing it"...

Comment (SmallJoker) [reply src/voxel.h:124]: Right. Such area also wouldn't fit in RAM so I suppose it's OK for now.

**Review by SmallJoker — APPROVED** [review]
Code looks good, the unittests do still pass and I couldn't find any new issue related to mapgen/vmanips.
